packaging/specifiers.py

```diff
@@ -119,7 +119,10 @@ def __str__(self) -> str:
 
     @property
     def _canonical_spec(self) -> Tuple[str, str]:
-        return self._spec[0], canonicalize_version(self._spec[1])
+        strip_zeros = self._spec[0] != "~="
+        return self._spec[0], canonicalize_version(
+            self._spec[1], strip_trailing_zero=strip_zeros
+        )
```

Comment:
How about this?

```suggestion
        canonical_version = canonicalize_version(
            self._spec[1],
            strip_trailing_zero=(self._spec[1] != "~="),
        )
        return self._spec[0], canonical_version
```

Because the zero-stripping logic is fine in general, just not for `~=`.